### DISCUSSION TITLE
How to measure the latency for each plugin?

### DISCUSSION THREAD

**Discussion by mong0520** [Kong/kong#8258] [Help] [answered]
Hi forks,

I have some self developed kong plugins enabled and I’d like to know if there is an easy way to measure/log each plugin’s latency for each request?

I am suffer from some random timeout issue and need to deep dive into the plugins.

**Comment by locao** [comment, marked as answer]
Hello @mong0520!

If you store the current time in the beginning of each phase handler and then calculate the time spent as the last thing in that phase handler you'll be able to log how much time was spent in each phase handler. Note that Nginx caches the time, so a call to `ngx.update_time` will give you a better time resolution: https://openresty-reference.readthedocs.io/en/latest/Lua_Nginx_API/#ngxnow

Please note that to calculate the whole plugin execution time you have to sum up the time spent in all phase handlers and not only store the time at the beginning of the first phase handler and calculate at the end of the last one, this would give you the time spent in other plugins and in Kong itself.
